web/modules/custom/joinup_group/src/Entity/PinnableGroupContentTrait.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\joinup_group\Entity;
+
+use Drupal\Core\Field\Plugin\Field\FieldType\EntityReferenceItem;
+use Drupal\Core\TypedData\Exception\MissingDataException;
+
+/**
+ * Reusable methods for pinnable group content entities.
+ *
+ * @todo Once we are on PHP 7.3 the JoinupBundleClassMetaEntityTrait
+ *   should be included here.
+ */
+trait PinnableGroupContentTrait {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isPinned(?GroupInterface $group = NULL): bool {
+    if (!$meta_entity = $this->getMetaEntity('pinned_in')) {
+      return FALSE;
+    }
+
+    /** @var \Drupal\joinup_federation\RdfEntityReferenceFieldItemList $item_list */
+    $item_list = $meta_entity->get('field_pinned_in');
+
+    if (empty($group)) {
+      return !$item_list->isEmpty();
+    }
+
+    return array_reduce($item_list->getValue(), function (bool $carry, array $properties) use ($group) {
+      return $carry || ($properties['target_id'] ?? '') === $group->id();
+    }, FALSE);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function pin(?GroupInterface $group = NULL): PinnableGroupContentInterface {
+    // Default to the parent group.
+    $group = $group ?: $this->getGroup();
+
+    // Only update the pinned status in the database if needed.
+    if (!$this->isPinned($group)) {
+      $meta_entity = $this->getMetaEntity('pinned_in');
+      if (empty($meta_entity)) {
+        $meta_entity = $this->createMetaEntity('pinned_in');
+      }
```

Comment:
I wanted to avoid the query that is done in `MetaEntity::loadOrCreate()` to get the entity ID. In both cases we don't need to do this query: when the meta entity already exists we can get the entity ID from the field data without having to query the database, and when it doesn't exist we know it and also don't need to do this query.

Reply:
Agree. I think we should change `::loadOrCreate()`, upstream, to do the same. However, there is some logic in doing the query: when the entity doesn't exist the query is cheaper than the loading.